### DISCUSSION TITLE
LED Positioning Indicator (LPI?)

### DISCUSSION THREAD

**Discussion by GalexY727** [Patribots4738/Crescendo2024#9] [Ideas]
`User Story:`
The starting area this year is much more open, and there isn't a strong incentive to start indexed against a wall if you could start further out to be closer to the pieces.

So, when we prepare our robot on the field, I think it would be useful to have an indicator to let the humans know if the robot is in the correct position for the chosen autonomous path. This would be very similar to what [Team 2363 did in 2023](https://www.youtube.com/watch?v=aRY7PFJ0PF0), except we don't plan on having an auto selector inside of the electronics of the robot. 

`Ideas:`
What we could do instead is compile a list of all possible starting configurations that the robot _could_ be at, and use the cameras to determine the closest one, and then have the LEDs tell us where we should move so that we can get instant feedback the ~~moment~~ a bit after the moment the robot turns on. 
- Come to think of it, it does still take one minute for the RoboRIO to boot up, we would have to turn it on maybe before we went on the field which doesn't sound that bad of an issue but it's worth mentioning.

**Comment by tina-hague** [comment]
Random question: Is it possible for the robot to also use its surroundings to know the closest part of the field it can end auto in? I remember you said we could have it deduct if it has a piece from the center and if it couldn’t pick it up, it will try to directly get the next piece instead of trying to shoot the one it doesn’t have. Would it be beneficial to do something similar for the end of auto so that the robot doesn’t try to get to the source and end up in the opponents wing when auto ends?

**Reply by GalexY727** [reply]
Because this feature would primarily effect the placement of the bot through the leds as "translators" from the code to the humans, it wouldn't be super useful to find the closest ending point of auto. I have been working a lot of paths recently, and we will be using event markers at preset positions to shoot the notes, which wouldn't necessarily be the end of an auto. Take a look at C3 S C2 for example, it starts at C3, comes back to shoot, then goes to pick up C2. The ending point is not the shooting point, the shooting point is merely just a pause in the middle of the path.

**Reply by GalexY727** [reply]
On the chance we go into the opponents wing, we would first need to find a way to track opposing robots and steer away from them so that we don't get penalized. There's a lot to do in auto, though, so we should be able to stay in our zone for 15 seconds.

**Comment by jacueblol** [comment]
Here are some resources I found for #13 

- https://docs.wpilib.org/en/stable/docs/software/hardware-apis/misc/addressable-leds.html
- https://www.chiefdelphi.com/t/some-questions-about-led-strips/400576
- https://store.ctr-electronics.com/1m-3-3ft-addressable-led-reel/
- https://www.amazon.com/s?k=WS2812+LEDs
- https://randomnerdtutorials.com/guide-for-ws2812b-addressable-rgb-led-strip-with-arduino/
- https://github.com/adafruit/Adafruit_NeoPixel
- https://store.ctr-electronics.com/candle/

NOTE: None of these are strictly necessary, and it is technically possible to wire these up with CAN.... Also, be careful about using any library or LED strip because it might overload your roborio.

**Reply by tina-hague** [reply]
This is everything I processed from your links: 
-there is a cool library on frc that works with the roborio
-there are many nice methods to use but periodic stuff is bad because much data?
-HSV is preferable
-there is a PWM port on the RoboRIO & the port # is what we input to address the LED
-there is a rainbow method! :D

**Comment by tina-hague** [comment]
[drawing board](https://www.tldraw.com/r/borcubmSklQQYMLikl6YZ?viewport=0,0,1296,791&page=page:page)

**Reply by GalexY727** [reply]
Can you add this to the drawing board corner in the pinned discussion post?

**Comment by jacueblol** [comment]
This is still in development, as we technically never got a chance to test out the less on the robot

**Reply by tina-hague** [reply]
This doesn’t need to be tested since it isn’t needed anymore. As long as our limelight has view of an April tag we can start auto anywhere in the starting zone… unless we somehow break both limelight’s at comp
